crates/neon/src/types_impl/buffer/types.rs

```diff
@@ -0,0 +1,760 @@
+
+impl JsArrayBuffer {
+    /// Constructs a new `JsArrayBuffer` object, safely zero-filled.
+    pub fn new<'a, C: Context<'a>>(cx: &mut C, len: usize) -> JsResult<'a, Self> {
+        let result = unsafe { sys::arraybuffer::new(cx.env().to_raw(), len) };
+
+        if let Ok(buf) = result {
+            Ok(Handle::new_internal(Self(buf)))
+        } else {
+            Err(Throw::new())
+        }
+    }
+
+    /// Construct a new `JsArrayBuffer` from bytes allocated by Rust
+    pub fn external<'a, C, T>(cx: &mut C, data: T) -> Handle<'a, Self>
+    where
+        C: Context<'a>,
+        T: AsMut<[u8]> + Send + 'static,
+    {
+        let env = cx.env().to_raw();
+        let value = unsafe { sys::arraybuffer::new_external(env, data) };
+
+        Handle::new_internal(Self(value))
+    }
+}
+
+unsafe impl TransparentNoCopyWrapper for JsArrayBuffer {
+    type Inner = raw::Local;
+
+    fn into_inner(self) -> Self::Inner {
+        self.0
+    }
+}
+
+impl Managed for JsArrayBuffer {
+    fn to_raw(&self) -> raw::Local {
+        self.0
+    }
+
+    fn from_raw(_env: Env, h: raw::Local) -> Self {
+        Self(h)
+    }
+}
+
+impl ValueInternal for JsArrayBuffer {
+    fn name() -> String {
+        "JsArrayBuffer".to_string()
+    }
+
+    fn is_typeof<Other: Value>(env: Env, other: &Other) -> bool {
+        unsafe { sys::tag::is_arraybuffer(env.to_raw(), other.to_raw()) }
+    }
+}
+
+impl Value for JsArrayBuffer {}
+
+impl Object for JsArrayBuffer {}
+
+impl private::Sealed for JsArrayBuffer {}
+
+impl TypedArray for JsArrayBuffer {
+    type Item = u8;
+
+    fn as_slice<'cx, 'a, C>(&self, cx: &'a C) -> &'a [Self::Item]
+    where
+        C: Context<'cx>,
+    {
+        unsafe { sys::arraybuffer::as_mut_slice(cx.env().to_raw(), self.to_raw()) }
+    }
+
+    fn as_mut_slice<'cx, 'a, C>(&mut self, cx: &'a mut C) -> &'a mut [Self::Item]
+    where
+        C: Context<'cx>,
+    {
+        unsafe { sys::arraybuffer::as_mut_slice(cx.env().to_raw(), self.to_raw()) }
+    }
+
+    fn try_borrow<'cx, 'a, C>(&self, lock: &'a Lock<C>) -> Result<Ref<'a, Self::Item>, BorrowError>
+    where
+        C: Context<'cx>,
+    {
+        // The borrowed data must be guarded by `Ledger` before returning
+        Ledger::try_borrow(&lock.ledger, unsafe {
+            sys::arraybuffer::as_mut_slice(lock.cx.env().to_raw(), self.to_raw())
+        })
+    }
+
+    fn try_borrow_mut<'cx, 'a, C>(
+        &mut self,
+        lock: &'a Lock<C>,
+    ) -> Result<RefMut<'a, Self::Item>, BorrowError>
+    where
+        C: Context<'cx>,
+    {
+        // The borrowed data must be guarded by `Ledger` before returning
+        Ledger::try_borrow_mut(&lock.ledger, unsafe {
+            sys::arraybuffer::as_mut_slice(lock.cx.env().to_raw(), self.to_raw())
+        })
+    }
+}
+
+/// A marker trait for all possible element types of binary buffers.
+///
+/// This trait can only be implemented within the Neon library.
+pub trait Binary: private::Sealed + Copy {
```

Comment:
I just pushed a change to make it pub-in-private. I think I like the docs better this way. It does still mention the `T: Binary` constraint on `JsTypedArray<T>` but with the big table in the `JsTypedArray` docs showing the full set of supported type arguments, I think it's fine.

Reply:
I had to revert this change because I realized that you can't write helper functions that operate generically on any `JsTypedArray<T>` unless you can refer to the `Binary` trait to add the proper constraint.